Build broken with XCode 16.3 (macOS SDK 15.4)

At least two issues:

## zlib

```
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/zlib/src/zutil.c:10:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/zlib/src/gzguts.h:20:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.4.sdk/usr/include/stdio.h:61:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.4.sdk/usr/include/_stdio.h:318:7: error: expected identifier or '('
  318 | FILE    *fdopen(int, const char *) __DARWIN_ALIAS_STARTING(__MAC_10_6, __IPHONE_2_0, __DARWIN_ALIAS(fdopen));
      |          ^
/Users/zwass/dev/osquery/libraries/cmake/source/zlib/src/zutil.h:147:33: note: expanded from macro 'fdopen'
  147 | #        define fdopen(fd,mode) NULL /* No fdopen() */
      |                                 ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/17/include/__stddef_null.h:26:16: note: expanded from macro 'NULL'
   26 | #define NULL ((void*)0)
      |                ^
```

This seems to be resolved by updating to zlib v1.3.1. See [forum thread](https://forums.bannister.org/ubbthreads.php?ubb=showflat&Number=124406).

## boost

```
In file included from /Users/zwass/dev/osquery/osquery/remote/http_client.cpp:11:
In file included from /Users/zwass/dev/osquery/build/ns_osquery_remote_httpclient/osquery/remote/http_client.h:31:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/deadline_timer.hpp:24:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/basic_deadline_timer.hpp:25:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/detail/deadline_timer_service.hpp:31:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/detail/timer_queue_ptime.hpp:22:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/time_traits.hpp:23:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/date_time/include/boost/date_time/posix_time/posix_time_types.hpp:16:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/date_time/include/boost/date_time/posix_time/posix_time_duration.hpp:15:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/cast.hpp:33:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/converter.hpp:13:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/conversion_traits.hpp:13:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/detail/conversion_traits.hpp:18:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/detail/int_float_mixture.hpp:19:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/mpl/include/boost/mpl/integral_c.hpp:32:
/Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/mpl/include/boost/mpl/aux_/integral_wrapper.hpp:73:31: error: integer value -1 is outside the valid range of values [0, 3] for the enumeration type 'udt_builtin_mixture_enum' [-Wenum-constexpr-conversion]
   73 |     typedef AUX_WRAPPER_INST( BOOST_MPL_AUX_STATIC_CAST(AUX_WRAPPER_VALUE_TYPE, (value - 1)) ) prior;
      |                               ^
/Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/mpl/include/boost/mpl/aux_/static_cast.hpp:24:47: note: expanded from macro 'BOOST_MPL_AUX_STATIC_CAST'
   24 | #   define BOOST_MPL_AUX_STATIC_CAST(T, expr) static_cast<T>(expr)
      |                                               ^
In file included from /Users/zwass/dev/osquery/osquery/remote/http_client.cpp:11:
In file included from /Users/zwass/dev/osquery/build/ns_osquery_remote_httpclient/osquery/remote/http_client.h:31:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/deadline_timer.hpp:24:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/basic_deadline_timer.hpp:25:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/detail/deadline_timer_service.hpp:31:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/detail/timer_queue_ptime.hpp:22:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/asio/include/boost/asio/time_traits.hpp:23:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/date_time/include/boost/date_time/posix_time/posix_time_types.hpp:16:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/date_time/include/boost/date_time/posix_time/posix_time_duration.hpp:15:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/cast.hpp:33:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/converter.hpp:13:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/conversion_traits.hpp:13:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/detail/conversion_traits.hpp:18:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/numeric/conversion/include/boost/numeric/conversion/detail/int_float_mixture.hpp:19:
In file included from /Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/mpl/include/boost/mpl/integral_c.hpp:32:
/Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/mpl/include/boost/mpl/aux_/integral_wrapper.hpp:73:31: error: integer value -1 is outside the valid range of values [0, 3] for the enumeration type 'int_float_mixture_enum' [-Wenum-constexpr-conversion]
/Users/zwass/dev/osquery/libraries/cmake/source/boost/src/libs/mpl/include/boost/mpl/aux_/static_cast.hpp:24:47: note: expanded from macro 'BOOST_MPL_AUX_STATIC_CAST'
   24 | #   define BOOST_MPL_AUX_STATIC_CAST(T, expr) static_cast<T>(expr)
      |                                               ^
2 errors generated.
make[2]: *** [osquery/remote/CMakeFiles/osquery_remote_httpclient.dir/http_client.cpp.o] Error 1
make[1]: *** [osquery/remote/CMakeFiles/osquery_remote_httpclient.dir/all] Error 2
```

Interestingly enough even with XCode 16.2 I still cannot build locally, I get this error, always from openssl:
```
In file included from apps/lib/app_libctx.c:9:
In file included from apps/include/app_libctx.h:13:
In file included from include/openssl/types.h:32:
include/openssl/e_os2.h:203:12: fatal error: 'sys/types.h' file not found
  203 | #  include <sys/types.h>
      |
```

> Interestingly enough even with XCode 16.2 I still cannot build locally, I get this error, always from openssl:
> 
> ```
> In file included from apps/lib/app_libctx.c:9:
> In file included from apps/include/app_libctx.h:13:
> In file included from include/openssl/types.h:32:
> include/openssl/e_os2.h:203:12: fatal error: 'sys/types.h' file not found
>   203 | #  include <sys/types.h>
>       |
> ```

Found the culprit, this was due to the recent CMake version I have vs what CI has, and the use of homebrew. The Note here explains it: https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_SYSROOT.html#variable:CMAKE_OSX_SYSROOT

> > Interestingly enough even with XCode 16.2 I still cannot build locally, I get this error, always from openssl:
> > ```
> > In file included from apps/lib/app_libctx.c:9:
> > In file included from apps/include/app_libctx.h:13:
> > In file included from include/openssl/types.h:32:
> > include/openssl/e_os2.h:203:12: fatal error: 'sys/types.h' file not found
> >   203 | #  include <sys/types.h>
> >       |
> > ```
> 
> Found the culprit, this was due to the recent CMake version I have (4.0+) vs what CI has, and the use of homebrew. The Note here explains it: https://cmake.org/cmake/help/latest/variable/CMAKE_OSX_SYSROOT.html#variable:CMAKE_OSX_SYSROOT

I had the same exact issue and I solved it by running:
```shell
xcrun --sdk macosx --show-sdk-path
```
Then used the output of the above command to configure cmake
```shell
cmake  -DCMAKE_OSX_SYSROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX15.5.sdk
```

During the boost upgrade I ran into some compatibility issues with https://github.com/osquery/osquery-toolchain
```
In file included from /usr/local/osquery-toolchain/usr/bin/../include/c++/v1/unordered_map:409:
/usr/local/osquery-toolchain/usr/bin/../include/c++/v1/__hash_table:2147:23: error: cannot initialize object parameter of type 'std::__1::__hash_node_base<std::__1::__hash_node<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, void *> *>' with an expression of type 'std::__1::__hash_node<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, void *>'
                    = __h.get()->__ptr();
                      ^~~~~~~~~
/usr/local/osquery-toolchain/usr/bin/../include/c++/v1/__hash_table:2154:16: error: static_cast from 'std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, void *>, std::__1::__hash_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, void *> > > >::pointer' (aka 'std::__1::__hash_node<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, void *> *') to 'std::__1::__hash_table<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, std::__1::__unordered_map_hasher<int, std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, std::__1::hash<int>, true>, std::__1::__unordered_map_equal<int, std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, std::__1::equal_to<int>, true>, std::__1::allocator<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor> > >::__next_pointer' (aka 'std::__1::__hash_node_base<std::__1::__hash_node<std::__1::__hash_value_type<int, osquery::ProcessContext::FileDescriptor>, void *> *> *'), which are not related by inheritance, is not allowed
        __nd = static_cast<__next_pointer>(__h.release());
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
14 errors generated.
make[2]: *** [osquery/events/CMakeFiles/osquery_events.dir/build.make:230: osquery/events/CMakeFiles/osquery_events.dir/linux/bpf/processcontextfactory.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 91%] Building CXX object plugins/config/parsers/CMakeFiles/plugins_config_parsers.dir/events_parser.cpp.o
make[1]: *** [CMakeFiles/Makefile2:17025: osquery/events/CMakeFiles/osquery_events.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Best guess is that we need to upgrade LLVM with compatible libc++

https://github.com/osquery/osquery/actions/runs/16142660515/job/45553603020?pr=8641

--- a/libraries/cmake/source/boost/CMakeLists.txt
+++ b/libraries/cmake/source/boost/CMakeLists.txt
@@ -146,6 +146,13 @@ function(importBoostInterfaceLibraries)
     BOOST_NO_CXX98_FUNCTION_BASE
   )
 
+  # silence enum constexpr conversion for MPL on Xcode 26
+  if(PLATFORM_MACOS)
+     target_compile_options(thirdparty_boost_mpl INTERFACE
+       -Wno-enum-constexpr-conversion
+      )
+  endif()
+
   # Additional settings for the libraries we just imported
   if(PLATFORM_LINUX)
     target_compile_definitions(thirdparty_boost_uuid INTERFACE

